### PR TITLE
Paper: Accelerating Scientific Python

### PR DESCRIPTION
**Pull request by anton-malakhov**
Title: Accelerating Scientific Python with Intel Optimizations

Authors: Oleksandr Pavlyk, Denis Nagorny, Andres Guzman-Ballen, Anton Malakhov

Keywords: numpy,scipy,sk-learn,numba,simd,parall,optimization,performance,FFT,random,MKL,DAAL,OpenMP,TBB,SVML

PLEASE NOTE: there are multiple .rst files included from the master file `optimizations.rst`. Github is unable to combine them into a single preview

### CODE REVIEW

**Comment by mpacer** [comment]
The current build system is not able to handle multiple `.rst` files — if you are leaving them there for reference for yourself putting them in a subdirectory should be sufficient for handling the reference use case. If you need both files to be part of your paper, that will require changing the paper building architecture. That kind of change is not going to be on the critical path for a little while since maintenance and bugfixes are occupying most of my attention.

**Comment by anton-malakhov** [comment]
@mpacer Does it mean `.. include::` tag is disabled? I relied on this doc when we were looking for a way to simplify synchronization between multiple contributes: http://docutils.sourceforge.net/docs/ref/rst/directives.html#including-an-external-document-fragment

**Comment by mpacer** [comment]
@anton-malakhov not necessarily, does putting them in a subdirectory work?

**Comment by anton-malakhov** [comment]
@mpacer Sure it works! We'll make appropriate changes soon. Thank you!

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+Fast Fourier Transforms
+-----------------------
+In addition to initial Fast Fourier Transforms (FFT) optimizations offered in previous releases, Update 2 brings widespread optimizations for NumPy and SciPy FFT. It offers a layered interface for the Intel |R| Math Kernel Library (Intel |R| MKL) that allows efficient access to native FFT optimizations from a range of NumPy and SciPy functions. The optimizations include real and complex data types, both single and double precision. Update 2 covers both one-dimensional and multi-dimensional data, as well as in place and out of place modes of operation. As a result, performance may improve up to 60x over Update 1 and is now close to performance of native C/Intel MKL.
+
+.. include:: papers/oleksandr_pavlyk/fft/fft.rst
+
+Arithmetic and transcendental expressions
+-----------------------------------------
+
+.. include:: papers/oleksandr_pavlyk/umath/umath.rst
```
Comment (anton-malakhov) [inline papers/oleksandr_pavlyk/optimizations.rst]: @mpacer It builds now but minor surprise is that it requires full path to the file starting from the root directory of the repository

**Comment by katyhuff** [comment]
@anton-malakhov I see that you have two posters related to "Accelerating Scientific Python". (Easy Steps for Accelerating Scientific Python and Secrets of Accelerating Scientific Python ). Can you please title your paper with exactly the same title as one of your two posters? Each paper should correspond to a single submitted abstract.

**Comment by anton-malakhov** [comment]
@katyhuff Yes, we submitted two abstracts for sake of telling the same story but from different angles. The first abstract was meant as a talk-only for regular users to show them how to use our optimizations (but was accepted as poster). The second one was supposed rather as a poster, mostly for developers and experts where we can discuss our performance results and ways to achieve them. The paper tries to do both things at the same time, thus the title consists of the common part. Though, it is leaning more towards the scope of the second abstract and we can go with its name if we cannot change its name anymore.

**Comment by katyhuff** [comment]
I see. It would be best if you can name it the same as the second abstract, then, if you're willing to do that. Thanks!

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+:video: http://www.youtube.com/watch?v=
+
+-------------------------------------------------------
+Accelerating Scientific Python with Intel Optimizations
+-------------------------------------------------------
+
+.. class:: abstract
+
+    It is well known that the performance difference between python and basic C code can be 200x.
```
Comment (serge-sans-paille) [inline papers/oleksandr_pavlyk/optimizations.rst]: This statement is misleading. When the Python code is written without the explicit loops, the performance improvement is rather x10 than x100.

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    Did you know that for numerically intensive code there is another 240x or more speedup possible?
```
Comment (serge-sans-paille) [inline papers/oleksandr_pavlyk/optimizations.rst]: I understand another 240x as 48000x which sounds surprising to me, *unless you exhibit some pathological test cases* but in that situation, better stress this up.

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+Faster Machine Learning with Scikit-learn
+-----------------------------------------
+
+.. include:: papers/oleksandr_pavlyk/sklearn/sklearn.rst
+
+Numba vectorization
+-------------------
+We worked with Continuum Analytics to make Numba to vectorize math code with transcedential functions using Intel SVML library.
+
```
Comment (serge-sans-paille) [inline papers/oleksandr_pavlyk/optimizations.rst]: Maybe you could quote http://iopscience.iop.org/article/10.1088/1749-4680/8/1/014001/meta here, as it also uses vector instruction (through boost.simd) for numerical python *disclaimer: i'm actually the pythran core dev*

File: papers/oleksandr_pavlyk/umath/umath.rst [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+SVML and this is thanks to VML's ability to both vectorize math functions and multi-thread loops.
+
+
+Specifically, on Intel |R| Core |R| i5 Intel Python delivers greater performance on transcendentals
+(log, exp, erf, etc.) due to utilizing both SIMD and multi-threading. We do not see any visible
+benefit of multi-threading basic operations (as shown on the graph) unless NumPy arrays are very
+large (not shown on the graph). On Xeon |R|, the 10x-1000x boost is explained by leveraging both
+(a) AVX2 instructions in transcendentals and (b) multiple cores (32 in our setup). Even greater
+scalability of Xeon Phi |R| relative to Xeon is explained by larger number of cores (64 in our
+setup) and a wider SIMD.
```
Comment (serge-sans-paille) [inline papers/oleksandr_pavlyk/umath/umath.rst]: maybe refering (and comparing to) numexpr here would be nice, as they also try to vectorize/parallelize numpy array expression.

File: papers/oleksandr_pavlyk/optimizations.rst (hunk 2 above)
Comment (anton-malakhov) [inline papers/oleksandr_pavlyk/optimizations.rst]: I guess this statement implies vectorization which automatically happens for some loops in C.
@oleksandr-pavlyk I guess we want to clarify here

File: papers/oleksandr_pavlyk/optimizations.rst (hunk 3 above)
Comment (anton-malakhov) [inline papers/oleksandr_pavlyk/optimizations.rst]: this is very possible with Xeon Phi as it has more than 240 threads. Though, sure, I agree, it needs to refer some prior work to claim such things in the abstract

File: papers/oleksandr_pavlyk/umath/umath.rst (hunk 5 above)
Comment (anton-malakhov) [inline papers/oleksandr_pavlyk/umath/umath.rst]: right, they do it pretty well, especially when linked against MKL - that's part of our distro anyway.

File: papers/oleksandr_pavlyk/optimizations.rst (hunk 1 above)
Comment (mpacer) [inline papers/oleksandr_pavlyk/optimizations.rst]: My guess is that has to do with how sphinx determines it's base path and relative paths to it. Honestly, I'm happy it's this uncomplicated :D.

**Comment by kapil-iitk** [comment]
I am unable to see your paper while syncing it is failed to build. Can you please upload the PDF of the file?

**Comment by mpacer** [comment]
There is a broken link in the paper to`sklearn_perf.jpg\'.

```
Error: File `sklearn_perf.jpg\' not found.\n\nSee the pdftex.def package documentation for explanation.
```

**Comment by mpacer** [comment]
@anton-malakhov As you seem to be taking the lead on this, I think you just need to change the content of that one link and this should build successfully. Comprehensive review deadlines are approaching and reviewers won't be able to reasonably review your paper if it cannot build.

**Comment by anton-malakhov** [comment]
@mpacer thanks for reminding! It is 'assembly' which is not recognized as a language name. fixed. it builds now

**Comment by anton-malakhov** [comment]
@kapil-iitk I apology that the paper is not meeting publishable state qualifications right now. We did our best to bring into this paper all the structure, thoughts, and facts we want to present but apparently it needs some more work. Out 'editing' author returns from vacation next week and will bring everything into coherent, plausible state. The paper builds now, BTW

**Comment by kapil-iitk** [comment]
I feel your paper require lot of modifications and clarifications.

1. The abstract of the paper should be concise and summery of the paper. Please rewrite the abstract.

2. The introduction must start with the issues and limitations of the available libraries, for which the intel optimizations are addressed.

3. "ACT OF PLAGIARISM" Most of the content of your paper is directly copied from the article "Intel Distribution for Python 2017 Update 2 accelerates five key areas for impressive performance gains" By Sergey Maidanov (Intel), Updated February 22, 2017 (Link: https://software.intel.com/en-us/articles/intelr-distribution-for-python-2017-update-2). I am arguing that this is not acceptable to use copyrighted material in the paper.

4. The size of the fonts embedded in the figures are too small. Please modify.

5. Abbreviations are discussed (even if they are well known). For example, in the abstract section, SIMD.

6. Table 3, must be shown in single column if it is horizontally not fit in the paper.

7. The code given in the paper exceeding the single column width, it is good to keep such code is the appendix and refer in the paper.

8. Please re-write the summary section based on the modification.

9. Citations and references are missing.

I hope these modification can improve the paper for the proceeding. I am suggesting to use latex/openoffice.org to prepare the manuscript for the review ( and to generate pdf).

**Comment by anton-malakhov** [comment]
Thank you, @kapil-iitk, for you feedback. I just want to quickly relax the concern regarding (3). Sergey is our manager and we a basically reuse our own text for sake of quickly filling this paper but I'm not arguing to keep this anyway. @oleksandr-pavlyk will rewrite this soon even though there is no obstacles for us to reuse this text.

File: papers/oleksandr_pavlyk/numba/svml.rst [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+where a computer program is converted from a scalar implementation, which processes a single pair of operands at a time,
+to a vector implementation, which processes one operation on multiple pairs of operands at once.
+
+According Numba's project page Numba is an Open Source NumPy-aware optimizing compiler for Python. 
+It uses the remarkable LLVM compiler infrastructure to compile Python syntax to machine code. And it quite expected that Numba tries
+to use all this features to improve performance especially for scientific applications. 
+
+
+LLVM implemented autovectorization for simple cases several years ago but there was quite sigificant problem with math functions.
+For proper vectorization it's needed to have special vectorized implementation of math fucntions as sin, cos, exp...
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/numba/svml.rst]: spelling: functions

File: papers/oleksandr_pavlyk/numba/svml.rst [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+According Numba's project page Numba is an Open Source NumPy-aware optimizing compiler for Python. 
+It uses the remarkable LLVM compiler infrastructure to compile Python syntax to machine code. And it quite expected that Numba tries
+to use all this features to improve performance especially for scientific applications. 
+
+
+LLVM implemented autovectorization for simple cases several years ago but there was quite sigificant problem with math functions.
+For proper vectorization it's needed to have special vectorized implementation of math fucntions as sin, cos, exp... 
+
+The Intel |R| C++ Compiler provides short vector math library (SVML) intrinsics to compute vector math functions.
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/numba/svml.rst]: The transition here is a bit weak

File: papers/oleksandr_pavlyk/numba/svml.rst [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+These intrinsics are available for IA-32 and Intel® 64 architectures running on supported operating systems.
+
+The SVML intrinsics are vector variants of corresponding scalar math operations using ``__m128``, ``__m128d``, ``__m256``, ``__m256d``, and ``__m256i`` data types.
+They take packed vector arguments, perform the operation on each element of the packed vector argument, and return a packed vector result.
+
+For example, the argument to the ``_mm_sin_ps`` intrinsic is a packed 128-bit vector of four 32-bit precision floating point numbers. The intrinsic computes the sine of each of these four numbers and returns the four results in a packed 128-bit vector.
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/numba/svml.rst]: Does packing incur an overhead that is still worthwhile?

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+:video: http://www.youtube.com/watch?v=
+
+-------------------------------------------------------
+Accelerating Scientific Python with Intel Optimizations
+-------------------------------------------------------
+
+.. class:: abstract
+
+    It is well known that the performance difference between Python and basic C code can be 200x.
+    Did you know that for numerically intensive code there is another 240x or more speedup possible?
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/optimizations.rst]: Rhetorical questions are probably best rephrased as statements: "can be 200x, and for numerically intensive code, the difference could be even greater."

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    The performance comes from software's ability to take advantage of your CPU's multiple cores,
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/optimizations.rst]: "Performance improvements derived from utilizing multiple cores, single instruction multiple data instructions, and high performance caches can acclerate Python programming to speeds previously limited to low-level languages like C."

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    single instruction multiple data (SIMD) instructions, and high performance caches.
+    This talk is for Python programmers who want to get the most out of their hardware
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/optimizations.rst]: You should rework the abstract to be for paper form and not the talk.

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    but do not have time or expertise to re-code their applications using techniques such as native
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/optimizations.rst]: What about dependencies that cannot be changed? Can those be accelerated as well?

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+   numpy,scipy,sk-learn,numba,simd,parallel,optimization,performance
+
+Introduction
+------------
+
+Intel released new version of its Distribution [IDP-Release]_. We suggest
+detailed report about optimization work we accomplished for Intel |R|
+Distribution for Python*, which might be interesting for developers of
+SciPy tools. It already offered great performance improvements for
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/optimizations.rst]: It has already offered...

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+Fast Fourier Transforms
+-----------------------
+
+In addition to initial optimizations of Fast Fourier Transform (FFT)
+offered in previous releases, Update 2 brings widespread optimizations
+for NumPy and SciPy FFT. It offers a think layered interface for the
+Intel |R| Math Kernel Library (Intel |R| MKL) that allows efficient
+access to native FFT optimizations from a range of NumPy and SciPy
+functions. The optimizations are provided for real and complex data
+types in both single and double precision. Update 2 improves
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/optimizations.rst]: "double precision floating point values".

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+Memory management optimizations
+-------------------------------
+
+Update 2 introduces widespread optimizations in NumPy memory
+management operations. As a dynamic language, Python manages memory
+for the user. Memory operations, such as allocation, de-allocation,
+copy, and move, affect performance of essentially all Python programs.
+Specifically, Update 2 ensures NumPy allocates arrays that are
+properly aligned in memory on Linux, so that NumPy and SciPy compute
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/optimizations.rst]: Aligned memory might be something that merits a little more explanation (or a diagram) for people less familiar with your domain.

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+functions can benefit from respective aligned versions of SIMD memory
+access instructions. This is especially relevant for Intel |R| Xeon
+Phi |TM| processors.  The most significant improvements in memory
+optimizations in Update 2 comes from replacing original memory copy
+and move operations with optimized implementations from Intel MKL. The
+result: improved performance because these Intel MKL routines are
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/optimizations.rst]: fragment

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+Auto-parallelization for Numba
+------------------------------
+
+.. include:: papers/oleksandr_pavlyk/parallel/parallel.rst
+
+
+Summary
+-------
+
+The Intel Distribution for Python is powered by Anaconda* and conda
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/optimizations.rst]: not powered by, but distributed with.

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+build infrastructures that give all Python users the benefit of
+interoperability within these two environments and access to the
+optimized packages through a simple conda install command.
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/optimizations.rst]: `conda install` should be in backticks.

File: papers/oleksandr_pavlyk/parallel/parallel.rst [hunk 19]
```diff
@@ -0,0 +1,78 @@
+In this section, we introduce a new feature in Numba that automatically parallelizes NumPy programs.
+Achieving high performance with Python on modern multi-core CPUs is challenging since Python implementations are generally interpreted and prohibit parallelism.
+To speed up sequential execution, Python functions can be compiled to fast native code using Numba, which in turn uses the LLVM JIT compiler.
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/parallel/parallel.rst]: remove "fast".

"using Numba, implemented with the LVVM JIT compiler"

should also define JIT (write out abbreviations on first time use) for those that are not familiar.

File: papers/oleksandr_pavlyk/parallel/parallel.rst [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+Achieving high performance with Python on modern multi-core CPUs is challenging since Python implementations are generally interpreted and prohibit parallelism.
+To speed up sequential execution, Python functions can be compiled to fast native code using Numba, which in turn uses the LLVM JIT compiler. 
+All a programmer has to do to use Numba is to annotate their functions with Numba's ``@jit`` decorator.
+However, the Numba JIT will not parallelize NumPy functions, even though the majority of them are known to have parallel semantics, and thus cannot make use of multiple cores.
+Furthermore, even if individual NumPy functions were parallelized, a program containing many such functions would likely have lackluster performance due to poor cache behavior. 
+Numba's existing solution is to allow users to write scalar kernels in OpenCL style, which can be executed in parallel. 
+However, this approach requires significant programming effort to rewrite existing array code into explicit parallelizable scalar kernels and therefore hurts productivity
+and is beyond the capabilities of some programmers.
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/parallel/parallel.rst]: why?

File: papers/oleksandr_pavlyk/sklearn/sklearn.rst [hunk 21]
```diff
@@ -0,0 +1,125 @@
+Scikit-learn is well know library that provides a lot of algorithms for many areas of machine learning.
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/sklearn/sklearn.rst]: well-known

File: papers/oleksandr_pavlyk/sklearn/sklearn.rst [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+
+        from sklearn.daal4sklearn import dispatcher
+        dispatcher.disable()
+        dispatcher.enable()
+
+We prepared several benchmarks to demonstrate performance that can be achieved with Intel |R| DAAL.
+
+.. code-block:: python
+
+        from __future__ import print_function
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/sklearn/sklearn.rst]: This example is quite large. A smaller example might be more fitting for the paper (maybe move this to an appendix)

File: papers/oleksandr_pavlyk/sklearn/sklearn.rst [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+        def train(X):
+            kmeans.fit(X)
+
+        for p, n in problem_sizes:
+            print (p,n, end=' ')
+            X_local = X[(p,n)]
+            train(X_local)
+            print('')
+
+Using all 32 cores of Intel |R| Xeon |R| processor E5-2698 v3 IDP’s K-Means can be more than 50 times faster than the python included with Ubuntu 14.04.
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/sklearn/sklearn.rst]: Showing relative speedup vs absolute speedup would be useful for those who are using commodity hardware, etc.

File: papers/oleksandr_pavlyk/umath/umath.rst [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+primitives, which are CPU-dispatched (to leverage appropriate ISA) and multi-threaded (leverage
+multiple cores) under the hood, and Intel |R| SVML intrinsics, a compiler-provided short vector
+math library that vectorizes math functions for both IA-32 and Intel |R| 64-bit architectures on
+supported operating systems. Depending on the problem size, NumPy will choose one of the two
+approaches. On much smaller array sizes, Intel |R| SVML outperforms VML due to VML's inherent cost
+of setting up the environment to multi-thread loops. For any other problem size, VML outperforms
+SVML and this is thanks to VML's ability to both vectorize math functions and multi-thread loops.
+
+
+Specifically, on Intel |R| Core |R| i5 Intel Python delivers greater performance on transcendentals
```
Comment (dalippa) [inline papers/oleksandr_pavlyk/umath/umath.rst]: A lot of parentheses used; it breaks up the paper flow. Consider using commas instead in certain places or reworking the sentences to flow better.

**Comment by anton-malakhov** [comment]
Hi @dalippa good to see you here again and thank you for the review! @oleksandr-pavlyk will address all the comments next week. I'm rather 'facilitating' author for this paper while I'm working on the second generation of composability paper (#282). If you are interested, I'll mention you for the next revision of that paper.

**Comment by kapil-iitk** [comment]
The paper is in good shape. There is still one thing need to be improved, text font within the figures are small.

**Comment by katyhuff** [comment]
Thank you for your reviews, @kapil-iitk @dalippa ! The authors seem to have made changes according to your comments. Do you feel this paper is now ready for inclusion in the SciPy proceedings?

**Pull request merged**